test/LiquidityManager/invest.test.ts

```diff
@@ -0,0 +1,489 @@
+                liquidityManager,
+            )
+    }
+
+    function getMinOutput(amount: bigint, tokenPrice: BigNumber) {
+        // Assume its stablecoin
+        if (tokenPrice.eq(USD_PRICE_BN))
+            return Slippage.deductSlippage(amount, SLIPPAGE_BN)
+
+        const amountBn = new BigNumber(amount.toString())
+
+        // Min output considering that we need to swap the token before
+        return Slippage.deductSlippage(
+            BigInt(amountBn.div(tokenPrice).toFixed(0)),
+            SLIPPAGE_BN.times(2),
+        )
+    }
+
+    async function invest(
+        pool: Pool,
+        swapToken0: string,
+        swapToken1: string,
+        priceToken0: BigNumber,
+        priceToken1: BigNumber,
+        {
+            swapAmountToken0,
+            swapAmountToken1,
+            tickLower,
+            tickUpper,
+        }: ReturnType<typeof UniswapV3.getMintPositionInfo>,
+    ) {
+        return liquidityManager
+            .connect(account0)
+            .investUniswapV3(
+                {
+                    positionManager: positionManagerUniV3,
+                    inputToken: stablecoin,
+                    depositAmountInputToken: amount,
+
+                    fee: pool.fee,
+
+                    token0: pool.token0.address,
+                    token1: pool.token1.address,
+
+                    swapToken0,
+                    swapToken1,
+
+                    swapAmountToken0,
+                    swapAmountToken1,
+
+                    tickLower,
+                    tickUpper,
+
+                    amount0Min: getMinOutput(swapAmountToken0, priceToken0),
+                    amount1Min: getMinOutput(swapAmountToken1, priceToken1),
+                },
+                permitAccount0,
+            )
+    }
+
+    function validateInvestTransaction(
+        receipt: ContractTransactionReceipt | null,
+        amount0: bigint,
+        token0Price: BigNumber,
+        amount1: bigint,
+        token1Price: BigNumber,
+    ) {
+        expect(receipt).to.be.an.instanceof(ContractTransactionReceipt)
+
+        const eventLog = getEventLog(receipt, 'Mint', UniswapV3Pool__factory.createInterface())
+
+        const mintedAmount0 = eventLog?.args.amount0
+        const mintedAmount1 = eventLog?.args.amount1
+
+        Compare.almostEqualPercentage({
+            target: amount0,
+            value: mintedAmount0,
+            tolerance: new BigNumber('0.1'),
+        })
+
+        Compare.almostEqualPercentage({
+            target: amount1,
+            value: mintedAmount1,
+            tolerance: new BigNumber('0.1'),
+        })
+
+        Compare.almostEqualPercentage({
+            target: amount,
+            value: BigInt(
+                new BigNumber(amount0.toString()).times(token0Price)
+                    .plus(new BigNumber(amount1.toString()).times(token1Price))
+                    .toString(),
+            ),
+            tolerance: new BigNumber('0.1'),
+        })
```

Comment:
10% tolerance is too much, i think it should be about 1-2%

Reply:
Oh its 10, i thought it was like 0.1% 🤣